### ISSUE TITLE
encounter `jit_flush_malloc_cache()` after hundreds of times of renderings

### ISSUE THREAD

**Issue by grkid** [mitsuba-renderer/mitsuba3#230]
I was trying to integrate mitsuba3 with pytorch. It goes on well until i meet this problem:

```
jit_flush_malloc_cache(): Dr.Jit exhausted the available memory and had to flush its allocation cache to free up additional memory. This is an expensive operation and will have a negative effect on performance. You may want to change your computation so that it uses less memory. This warning will only be displayed once.


Critical Dr.Jit compiler failure: cuda_check(): API error 0700 (CUDA_ERROR_ILLEGAL_ADDRESS): "an illegal memory access was encountered" in D:\a\drjit\drjit\ext\drjit-core\src\malloc.cpp:377.
```

It ALWAYS appears after hundreds of times of (`cuda_ad_rgb`) renderings. Seems like the memory on GPU has run out.

More Infomation:
1. I render the same scene with different params about 300+ times (for gradient information) and it appears.
2. I use a single RTX 3090 graphic card, with 24GB on-board memory.
3. the rendered image's resolution is 256*256, and spp=4.
4. I integrate mitsuba3+drjit with pytorch, and both of them uses CUDA.
5. The Integrator of the scene is `prbvolpath`.
6. Windows 10 + 5800X CPU, CUDA 11.6, nvidia driver version 512.15.
7. Sometimes the gradient can be `NaN`. But it won't appear often.

Any assistance will be appreciated.

**Comment by grkid** [comment]
Maybe switch to 'llvm_ad_rgb' can solve this problem?

**Comment by Speierers** [comment]
Hi @grkid, do you evaluate the parameter's gradient after every render call? Otherwise it might be that the JIT compiler keeps some internal state alive in between iteration, which will eat up all the memory available on your GPU.

Could you maybe share a simplified version of your pipeline?

**Comment by grkid** [comment]
```python
class MitsubaLossFunction(torch.autograd.Function):

    def __init__(self):
        super(MitsubaLossFunction, self).__init__()

    @staticmethod
    def forward(ctx, net_output, ref_output, ref_image, scene, params):

        key_sigma_t = 'juice2.sigma_t.value.value'
        key_albedo = 'juice2.albedo.value.value'
        key_g = 'hg1.g'

        output_numpy = net_output[0:7].cpu().detach().numpy()
        sigma_t = output_numpy[0:3]
        albedo = output_numpy[3:6]
        g = output_numpy[6:7]

        sigma_t = np.clip(sigma_t, 0.0, 1.0)
        albedo = np.clip(albedo, 0.0, 1.0)
        g = np.clip(g, 0.0, 1.0)
        params[key_sigma_t] = mi.Color3f(sigma_t)
        params[key_albedo] = mi.Color3f(albedo)
        params[key_g] = float(g)
        params.update()

        dr.enable_grad(params[key_sigma_t])
        dr.enable_grad(params[key_albedo])
        dr.enable_grad(params[key_g])
        params.update()

        image = mi.render(scene, params, spp=RENDER_SPP)

        mse_loss = dr.mean(dr.sqr(image - ref_image))

        dr.backward(mse_loss)
        grad_loss_sigma_t = np.array(dr.grad(params[key_sigma_t]))  # d(Loss)/d(sigma_t)
        grad_loss_albedo = np.array(dr.grad(params[key_albedo]))  # d(Loss)/d(albedo)
        grad_loss_g = np.array(dr.grad(params[key_g]))  # d(Loss)/d(g)

        ctx.save_for_backward(net_output)
        ctx.grad_loss_sigma_t = grad_loss_sigma_t
        ctx.grad_loss_albedo = grad_loss_albedo
        ctx.grad_loss_g = grad_loss_g

        loss = torch.Tensor(1, 1)
        numpyloss = mse_loss.numpy()
        loss[0][0] = torch.from_numpy(numpyloss)

        return loss

    @staticmethod
    def backward(ctx, grad_output):
        net_output, = ctx.saved_tensors
        grad_loss_sigma_t = ctx.grad_loss_sigma_t
        grad_loss_albedo = ctx.grad_loss_albedo
        grad_loss_g = ctx.grad_loss_g
        grad_in = torch.Tensor(net_output.size()).float().to(device)
        grad_in[0:3] = grad_output.clone()[0][0] * grad_loss_sigma_t
        grad_in[3:6] = grad_output.clone()[0][0] * grad_loss_albedo
        grad_in[6:7] = grad_output.clone()[0][0] * grad_loss_g
        return grad_in, None, None, None, None

def train():
    for epoch in range(num_epoches):
        running_loss = 0.0
        for batch_idx, data in enumerate(train_loader):
            lossMTS = 0

            inputs, target, origin, fn = ...

            target = target.cpu().numpy().tolist()
            outputs = model(inputs)
            outputs.retain_grad()
            optimizer.zero_grad()

            tensortarget = torch.tensor(target).to(device)
            # tensortarget = torch.clamp(tensortarget, 0, 0.9999)
            losspara = criterion(outputs, tensortarget)

            for i in range(len(target)):
                currentFn=fn[i]
                print("i: ", i, " outputs: ", outputs[i], " target: ", tensortarget[i])

     
                image_ref = mitsubaImageDict[currentFn]
                ref_numpy = target[i]
                lossMTS = lossMTS + criterion_MTS(outputs[i], ref_numpy, image_ref, scene_criterion, criterion_params)
               

            # after every batch , compute loss and backpropagate to the net
            loss = IMG_LOSS_RATIO*lossMTS.to(device) + PARAM_LOSS_RATIO*losspara
            loss.requires_grad_(True)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
```

`model` is a standard torch network.

**Comment by grkid** [comment]
`criterion_MTS` is `MitsubaLossFunction.apply`.

**Comment by grkid** [comment]
Besides, i don't actually understand the meaning of 'evaluate'. Seems like it has nothing to do with garbage collection?

**Comment by Speierers** [comment]
Looking at your code, I don't see why Mitsuba or Dr.Jit would use more memory as we increase the iteration count. I suspect it is PyTorch that consumes most of the GPU memory, to the point where Mitsuba / DrJit cannot allocate the necessary resources to perform the rendering. 

For this I suggest you make sure that the variables are properly garbage collected after every iteration of your optimization loop.

**Comment by grkid** [comment]
After modifying some unrelated code, it happens no more. I will spend some time to find how and why.
